File: tests/python/test_tree.py
# Gingo — Tree tests (Python)
# SPDX-License-Identifier: MIT

import pytest
from gingo import Tree, ScaleType, HarmonicFunction


# ---------------------------------------------------------------------------
# Construction
# ---------------------------------------------------------------------------

def test_tree_construction_from_scale_type():
    t = Tree("C", ScaleType.Major)
    assert t.tonic().natural() == "C"
    assert t.type() == ScaleType.Major


def test_tree_construction_from_string():
    t = Tree("C", "major")
    assert t.tonic().natural() == "C"
    assert t.type() == ScaleType.Major


def test_tree_construction_different_tonics():
    t1 = Tree("G", ScaleType.Major)
    assert t1.tonic().natural() == "G"

    t2 = Tree("A", ScaleType.NaturalMinor)
    assert t2.tonic().natural() == "A"


# ---------------------------------------------------------------------------
# Branches
# ---------------------------------------------------------------------------

def test_tree_branches_non_empty():
    t = Tree("C", ScaleType.Major)
    branches = t.branches()
    assert len(branches) > 0


def test_tree_branches_major_scale():
    t = Tree("C", ScaleType.Major)
    branches = t.branches()

    # Should contain basic diatonic branches
    assert "I" in branches
    assert "V7" in branches
    assert "IIm" in branches


def test_tree_branches_minor_scale():
    t = Tree("A", ScaleType.NaturalMinor)
    branches = t.branches()

    # Should contain Im
    assert "Im" in branches


# ---------------------------------------------------------------------------
# Paths
# ---------------------------------------------------------------------------

def test_tree_paths_from_origin():
    t = Tree("C", ScaleType.Major)
    paths = t.paths("I")

    assert len(paths) > 0
    # First path should be the origin itself
    assert paths[0].branch == "I"
    assert paths[0].id == 0


def test_tree_paths_contain_chord_info():
    t = Tree("C", ScaleType.Major)
    paths = t.paths("I")

    # Origin should have chord resolved
    assert paths[0].chord.root().natural() == "C"
    assert len(paths[0].interval_labels) > 0
    assert len(paths[0].note_names) > 0


# ---------------------------------------------------------------------------
# Shortest path
# ---------------------------------------------------------------------------

def test_tree_shortest_path_direct():
    t = Tree("C", ScaleType.Major)
    path = t.shortest_path("I", "V7")

    assert len(path) > 0
    assert path[0] == "I"
    assert path[-1] == "V7"


def test_tree_shortest_path_invalid_branch():
    t = Tree("C", ScaleType.Major)
    path = t.shortest_path("INVALID", "V7")

    assert len(path) == 0


def test_tree_shortest_path_same_branch():
    t = Tree("C", ScaleType.Major)
    path = t.shortest_path("I", "I")

    assert len(path) == 1
    assert path[0] == "I"


def test_tree_shortest_path_multi_step():
    t = Tree("C", ScaleType.Major)
    path = t.shortest_path("I", "IV")

    # Should find some path, even if indirect
    if len(path) > 0:
        assert path[0] == "I"
        assert path[-1] == "IV"
        assert len(path) >= 2


# ---------------------------------------------------------------------------
# Valid progression
# ---------------------------------------------------------------------------

def test_tree_valid_progression_single():
    t = Tree("C", ScaleType.Major)
    assert t.is_valid_progression(["I"])


def test_tree_valid_progression_ii_v_i():
    t = Tree("C", ScaleType.Major)
    # Classic II-V-I should be valid if paths exist
    valid = t.is_valid_progression(["IIm", "V7", "I"])
    # Don't require it to be true, just check it doesn't crash
    assert isinstance(valid, bool)


def test_tree_valid_progression_empty():
    t = Tree("C", ScaleType.Major)
    assert t.is_valid_progression([])


def test_tree_valid_progression_invalid_branch():
    t = Tree("C", ScaleType.Major)
    # Test with progression containing invalid branch
    # Single invalid branch may return True as empty path is valid
    # Test with progression that has invalid transition
    assert not t.is_valid_progression(["I", "INVALID", "V7"])


# ---------------------------------------------------------------------------
# Harmonic function
# ---------------------------------------------------------------------------

def test_tree_function_tonic():
    t = Tree("C", ScaleType.Major)
    func = t.function("I")
    assert func == HarmonicFunction.Tonic


def test_tree_function_subdominant():
    t = Tree("C", ScaleType.Major)
    func = t.function("IV")
    assert func == HarmonicFunction.Subdominant


def test_tree_function_dominant():
    t = Tree("C", ScaleType.Major)
    func = t.function("V7")
    assert func == HarmonicFunction.Dominant


def test_tree_function_applied_uses_primary():
    t = Tree("C", ScaleType.Major)
    # V7/IV should be Dominant (V7 = Dominant)
    func = t.function("V7 / IV")
    assert func == HarmonicFunction.Dominant


def test_tree_function_minor_branches():
    t = Tree("C", ScaleType.Major)

    func_iim = t.function("IIm")
    assert func_iim == HarmonicFunction.Subdominant

    func_vim = t.function("VIm")
    assert func_vim == HarmonicFunction.Tonic


# ---------------------------------------------------------------------------
# Branches with function
# ---------------------------------------------------------------------------

def test_tree_branches_with_function_tonics():
    t = Tree("C", ScaleType.Major)
    tonics = t.branches_with_function(HarmonicFunction.Tonic)

    assert len(tonics) > 0
    # Should contain I
    assert "I" in tonics


def test_tree_branches_with_function_subdominants():
    t = Tree("C", ScaleType.Major)
    subdominants = t.branches_with_function(HarmonicFunction.Subdominant)

    assert len(subdominants) > 0


def test_tree_branches_with_function_dominants():
    t = Tree("C", ScaleType.Major)
    dominants = t.branches_with_function(HarmonicFunction.Dominant)

    assert len(dominants) > 0
    # Should contain V7
    assert "V7" in dominants


# ---------------------------------------------------------------------------
# Visualization export
# ---------------------------------------------------------------------------

def test_tree_to_dot_basic():
    t = Tree("C", ScaleType.Major)
    dot = t.to_dot(False)

    assert len(dot) > 0
    assert "digraph" in dot


def test_tree_to_dot_with_functions():
    t = Tree("C", ScaleType.Major)
    dot = t.to_dot(True)

    assert len(dot) > 0
    # Should have color information
    assert "fillcolor" in dot


def test_tree_to_mermaid():
    t = Tree("C", ScaleType.Major)
    mermaid = t.to_mermaid()

    assert len(mermaid) > 0
    assert "graph" in mermaid


# ---------------------------------------------------------------------------
# Diminished chord notation
# ---------------------------------------------------------------------------

def test_tree_diminished_notation():
    t = Tree("C", ScaleType.Major)
    all_branches = t.branches()

    # Check if diminished notation exists (either dim or °)
    has_dim = any("dim" in b or "°" in b for b in all_branches)
    # Should support at least one notation
    assert has_dim


# ---------------------------------------------------------------------------
# Different scale types
# ---------------------------------------------------------------------------

def test_tree_natural_minor():
    t = Tree("A", ScaleType.NaturalMinor)
    branches = t.branches()
    assert len(branches) > 0

    paths = t.paths("Im")
    assert len(paths) > 0


def test_tree_harmonic_minor():
    t = Tree("A", ScaleType.HarmonicMinor)
    branches = t.branches()
    assert len(branches) > 0


def test_tree_melodic_minor():
    t = Tree("A", ScaleType.MelodicMinor)
    branches = t.branches()
    assert len(branches) > 0


# ---------------------------------------------------------------------------
# String representation
# ---------------------------------------------------------------------------

def test_tree_to_string():
    t = Tree("C", ScaleType.Major)
    s = str(t)

    assert len(s) > 0
    assert "C" in s


def test_tree_repr():
    t = Tree("C", ScaleType.Major)
    r = repr(t)

    assert len(r) > 0

File: tests/cpp/test_tree.cpp
// Gingo — Tree module tests
// SPDX-License-Identifier: MIT

#include <catch2/catch_test_macros.hpp>
#include <gingo/tree.hpp>

using namespace gingo;

// ---------------------------------------------------------------------------
// Construction
// ---------------------------------------------------------------------------

TEST_CASE("Tree construction from ScaleType", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    REQUIRE(t.tonic().natural() == "C");
    REQUIRE(t.type() == ScaleType::Major);
}

TEST_CASE("Tree construction from string", "[tree]") {
    Tree t("C", "major", "harmonic_tree");
    REQUIRE(t.tonic().natural() == "C");
    REQUIRE(t.type() == ScaleType::Major);
}

TEST_CASE("Tree construction with different tonics", "[tree]") {
    Tree t1("G", ScaleType::Major, "harmonic_tree");
    REQUIRE(t1.tonic().natural() == "G");

    Tree t2("A", ScaleType::NaturalMinor, "harmonic_tree");
    REQUIRE(t2.tonic().natural() == "A");
}

// ---------------------------------------------------------------------------
// Branches
// ---------------------------------------------------------------------------

TEST_CASE("Tree branches returns non-empty list", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto branches = t.branches();
    REQUIRE(!branches.empty());
}

TEST_CASE("Tree branches for major scale", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto branches = t.branches();

    // Should contain basic diatonic branches
    bool has_I = false;
    bool has_V7 = false;
    bool has_IIm = false;

    for (const auto& b : branches) {
        if (b == "I") has_I = true;
        if (b == "V7") has_V7 = true;
        if (b == "IIm") has_IIm = true;
    }

    REQUIRE(has_I);
    REQUIRE(has_V7);
    REQUIRE(has_IIm);
}

TEST_CASE("Tree branches for minor scale", "[tree]") {
    Tree t("A", ScaleType::NaturalMinor, "harmonic_tree");
    auto branches = t.branches();

    // Should contain Im
    bool has_Im = false;
    for (const auto& b : branches) {
        if (b == "Im") has_Im = true;
    }
    REQUIRE(has_Im);
}

// ---------------------------------------------------------------------------
// Paths
// ---------------------------------------------------------------------------

TEST_CASE("Tree paths from origin", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto paths = t.paths("I");

    REQUIRE(!paths.empty());
    // First path should be the origin itself
    REQUIRE(paths[0].branch == "I");
    REQUIRE(paths[0].id == 0);
}

TEST_CASE("Tree paths contain chord information", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto paths = t.paths("I");

    // Origin should have chord resolved
    REQUIRE(paths[0].chord.root().natural() == "C");
    REQUIRE(!paths[0].interval_labels.empty());
    REQUIRE(!paths[0].note_names.empty());
}

// ---------------------------------------------------------------------------
// Shortest path
// ---------------------------------------------------------------------------

TEST_CASE("Tree shortest_path finds direct connection", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto path = t.shortest_path("I", "V7");

    REQUIRE(!path.empty());
    REQUIRE(path.front() == "I");
    REQUIRE(path.back() == "V7");
}

TEST_CASE("Tree shortest_path returns empty for invalid branches", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto path = t.shortest_path("INVALID", "V7");

    REQUIRE(path.empty());
}

TEST_CASE("Tree shortest_path from branch to itself", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto path = t.shortest_path("I", "I");

    REQUIRE(path.size() == 1);
    REQUIRE(path[0] == "I");
}

TEST_CASE("Tree shortest_path finds multi-step path", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto path = t.shortest_path("I", "IV");

    // Should find some path, even if indirect
    if (!path.empty()) {
        REQUIRE(path.front() == "I");
        REQUIRE(path.back() == "IV");
        REQUIRE(path.size() >= 2);
    }
}

// ---------------------------------------------------------------------------
// Valid progression
// ---------------------------------------------------------------------------

TEST_CASE("Tree is_valid_progression single chord", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    REQUIRE(t.is_valid({"I"}));
}

TEST_CASE("Tree is_valid_progression II-V-I", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    // Classic II-V-I should be valid if paths exist
    auto valid = t.is_valid({"IIm", "V7", "I"});
    // Don't require it to be true, just check it doesn't crash
    REQUIRE((valid == true || valid == false));
}

TEST_CASE("Tree is_valid_progression empty", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    REQUIRE(t.is_valid({}));
}

TEST_CASE("Tree is_valid_progression with invalid branch", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    // Test progression containing invalid branch
    REQUIRE_FALSE(t.is_valid({"I", "INVALID", "V7"}));
}

// ---------------------------------------------------------------------------
// Harmonic function
// ---------------------------------------------------------------------------

TEST_CASE("Tree function returns tonic for I", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto func = t.function("I");
    REQUIRE(func == HarmonicFunction::Tonic);
}

TEST_CASE("Tree function returns subdominant for IV", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto func = t.function("IV");
    REQUIRE(func == HarmonicFunction::Subdominant);
}

TEST_CASE("Tree function returns dominant for V7", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto func = t.function("V7");
    REQUIRE(func == HarmonicFunction::Dominant);
}

TEST_CASE("Tree function for applied chords uses primary degree", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    // V7/IV should be Dominant (V7 = Dominant)
    auto func = t.function("V7 / IV");
    REQUIRE(func == HarmonicFunction::Dominant);
}

TEST_CASE("Tree function for minor branches", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto func_IIm = t.function("IIm");
    REQUIRE(func_IIm == HarmonicFunction::Subdominant);

    auto func_VIm = t.function("VIm");
    REQUIRE(func_VIm == HarmonicFunction::Tonic);
}

// ---------------------------------------------------------------------------
// Branches with function
// ---------------------------------------------------------------------------

TEST_CASE("Tree branches_with_function returns tonics", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto tonics = t.branches_with_function(HarmonicFunction::Tonic);

    REQUIRE(!tonics.empty());
    // Should contain I
    bool has_I = false;
    for (const auto& b : tonics) {
        if (b == "I") has_I = true;
    }
    REQUIRE(has_I);
}

TEST_CASE("Tree branches_with_function returns subdominants", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto subdominants = t.branches_with_function(HarmonicFunction::Subdominant);

    REQUIRE(!subdominants.empty());
}

TEST_CASE("Tree branches_with_function returns dominants", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto dominants = t.branches_with_function(HarmonicFunction::Dominant);

    REQUIRE(!dominants.empty());
    // Should contain V7
    bool has_V7 = false;
    for (const auto& b : dominants) {
        if (b == "V7") has_V7 = true;
    }
    REQUIRE(has_V7);
}

// ---------------------------------------------------------------------------
// Visualization export
// ---------------------------------------------------------------------------

TEST_CASE("Tree to_dot returns non-empty string", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto dot = t.to_dot(false);

    REQUIRE(!dot.empty());
    REQUIRE(dot.find("digraph") != std::string::npos);
}

TEST_CASE("Tree to_dot with functions includes colors", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto dot = t.to_dot(true);

    REQUIRE(!dot.empty());
    // Should have color information
    REQUIRE(dot.find("fillcolor") != std::string::npos);
}

TEST_CASE("Tree to_mermaid returns non-empty string", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto mermaid = t.to_mermaid();

    REQUIRE(!mermaid.empty());
    REQUIRE(mermaid.find("graph") != std::string::npos);
}

// ---------------------------------------------------------------------------
// Diminished chord notation (both ° and dim)
// ---------------------------------------------------------------------------

TEST_CASE("Tree resolves diminished chord notation", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto paths = t.paths("I");

    // Check if Idim or I° exists in branches (both ° and dim notations)
    auto all_branches = t.branches();

    bool has_dim = false;
    for (const auto& b : all_branches) {
        if (b == "Idim" || b.find("°") != std::string::npos) {
            has_dim = true;
            break;
        }
    }
    // Should support at least one diminished notation (dim or °)
    REQUIRE(has_dim);
}

// ---------------------------------------------------------------------------
// Integration with different scale types
// ---------------------------------------------------------------------------

TEST_CASE("Tree works with NaturalMinor", "[tree]") {
    Tree t("A", ScaleType::NaturalMinor, "harmonic_tree");
    auto branches = t.branches();
    REQUIRE(!branches.empty());

    auto paths = t.paths("Im");
    REQUIRE(!paths.empty());
}

TEST_CASE("Tree works with HarmonicMinor", "[tree]") {
    Tree t("A", ScaleType::HarmonicMinor, "harmonic_tree");
    auto branches = t.branches();
    REQUIRE(!branches.empty());
}

TEST_CASE("Tree works with MelodicMinor", "[tree]") {
    Tree t("A", ScaleType::MelodicMinor, "harmonic_tree");
    auto branches = t.branches();
    REQUIRE(!branches.empty());
}

// ---------------------------------------------------------------------------
// to_string
// ---------------------------------------------------------------------------

TEST_CASE("Tree to_string includes tonic and type", "[tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto str = t.to_string();

    REQUIRE(!str.empty());
    REQUIRE(str.find("C") != std::string::npos);
}
